File: LeetCode/Study Plan/Practice Problems/Medium/Design Browser History/Go/designBrowserHistory.go
package main

import "fmt"

type BrowserHistory struct {
	stack   []string
	pointer int
}

func Constructor(homepage string) BrowserHistory {
	st := []string{homepage}
	return BrowserHistory{stack: st, pointer: 1}
}

func (this *BrowserHistory) Visit(url string) {
	this.stack = this.stack[:this.pointer]
	this.stack = append(this.stack, url)
	this.pointer++
	fmt.Println(this.stack, this.pointer)
}

func (this *BrowserHistory) Back(steps int) string {
	if this.pointer <= steps {
		this.pointer = 1
		return this.stack[0]
	}
	this.pointer -= steps
	return this.stack[this.pointer-1]
}

func (this *BrowserHistory) Forward(steps int) string {
	if this.pointer+steps >= len(this.stack) {
		this.pointer = len(this.stack)
		return this.stack[this.pointer-1]
	}
	this.pointer += steps
	return this.stack[this.pointer-1]
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * obj := Constructor(homepage);
 * obj.Visit(url);
 * param_2 := obj.Back(steps);
 * param_3 := obj.Forward(steps);
 */

File: LeetCode/Medium/Design Browser History/Java/designBrowserHistory.java
class BrowserHistory {
    ArrayList<String> urls;
    int currPointer;
    int lastPointer;

    public BrowserHistory(String homepage) {
        urls = new ArrayList<>();
        urls.add(homepage);
        currPointer = 0;
        lastPointer = 0;    
    }
    
    public void visit(String url) {
        currPointer += 1;
        if (urls.size() > currPointer) urls.set(currPointer, url);
        else urls.add(url);
        lastPointer = currPointer;
    }
    
    public String back(int steps) {
        currPointer = Math.max(currPointer-steps, 0);
        return urls.get(currPointer);
    }
    
    public String forward(int steps) {
        currPointer = Math.min(lastPointer, currPointer+steps);
        return urls.get(currPointer);
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */
